### PR TITLE
Sprite: basic animation support

### PR DESCRIPTION
**Pull request by moneromooo**

### CODE REVIEW

File: Source/Urho3D/UI/Sprite.cpp [hunk 1]
```diff
@@ -206,8 +225,16 @@ void Sprite::SetImageRect(const IntRect& rect)
...
-    if (texture_)
-        SetImageRect(IntRect(0, 0, texture_->GetWidth(), texture_->GetHeight()));
+    if (!textures_.Empty())
+    {
+        int width = 0, height = 0;
+        for (const auto &texture: textures_)
+        {
+            width = std::max(width, texture.first_->GetWidth());
+            height = std::max(height, texture.first_->GetHeight());
+            SetImageRect(IntRect(0, 0, width, height));
```
Comment (UkoeHB) [inline Source/Urho3D/UI/Sprite.cpp:235]: Seems like `SetImageRect` should be outside the loop

File: Source/Urho3D/UI/Sprite.h [hunk 2]
```diff
@@ -117,13 +125,16 @@ class URHO3D_API Sprite : public UIElement
     /// Rotation angle.
     float rotation_;
     /// Texture.
-    SharedPtr<Texture> texture_;
+    Vector<Pair<SharedPtr<Texture>, unsigned int>> textures_;
```
Comment (UkoeHB) [inline Source/Urho3D/UI/Sprite.h:128]: Comment out of date

File: Source/Urho3D/UI/Sprite.h [hunk 3]
```diff
@@ -72,6 +72,8 @@ class URHO3D_API Sprite : public UIElement
     void SetRotation(float angle);
     /// Set texture.
     void SetTexture(Texture* texture);
+    /// Set animation.
+    void SetAnimation(const Vector<Pair<SharedPtr<Texture>, unsigned int>> &textures);
```
Comment (UkoeHB) [inline Source/Urho3D/UI/Sprite.h:76]: Maybe clarify the comment here about what the first parameter is. Seems like a set of frames, where the second part of the pair indicates approximately how long (in milliseconds?) the frame should be active before the next frame can appear.

File: Source/Urho3D/UI/Sprite.cpp (hunk 1 above)
Comment (UkoeHB) [reply Source/Urho3D/UI/Sprite.cpp:235]: Also maybe clarify intent
```
// set sprite image rectangle so it fits all frame textures
```

File: Source/Urho3D/UI/Sprite.cpp [hunk 4]
```diff
@@ -289,7 +316,38 @@ void Sprite::SetTextureAttr(const ResourceRef& value)
...
+
+    unsigned ms = timer.GetMSec(false);
+    unsigned t = ms % total_time;
+    frame_ = 0;
+    for (const auto &texture: textures_)
+    {
+        if (t < texture.second_)
+            break;
+        t -= texture.second_;
+        ++frame_;
```
Comment (UkoeHB) [inline Source/Urho3D/UI/Sprite.cpp:349]: Clarify intent, it's hard to understand this at first glance
```
// get the total time it takes to cycle through all frames
unsigned total_time = 0;
...
// if frames don't have timing information, then we just iterate through them with every update
if (total_time == 0)
...
// how far into this 'animation cycle' we are, based on time elapsed
unsigned ms = timer.GetMSec(false);
...
// figure out which animation frame this position in the cycle corresponds to
for (const auto &texture: textures_)
...
```

**Comment by eugeneko** [comment]
I didn't check this code in-depth, but I see one critical issue so far:
This animation cannot be serialized, i.e. it cannot be configured from style file or from UI Editor.

Also, this functionality overlaps a bit with `ValueAnimation` that can animate everything, not just textures.
Can this code do anything that `ValueAnimation` cannot?

**Comment by moneromooo** [comment]
No idea, I did not see ValueAnimation. There seems to be no point in this then.

**Pull request closed without merging**